R/susie_engine.R

```diff
@@ -0,0 +1,125 @@
+                        L, intercept = TRUE, standardize = TRUE,
+                        scaled_prior_variance, residual_variance,
+                        prior_weights, null_weight, model_init = NULL,
+                        estimate_prior_variance,
+                        estimate_prior_method = c("optim", "EM", "simple"),
+                        check_null_threshold,
+                        estimate_residual_variance,
+                        residual_variance_lowerbound,
+                        residual_variance_upperbound,
+                        max_iter, tol, verbose, track_fit,
+                        coverage, min_abs_corr,
+                        prior_tol, n_purity, compute_univariate_zscore = FALSE,
+                        check_prior = FALSE,
+                        refine = FALSE){
+
+  # Prior Variance Method
+  estimate_prior_method <- match.arg(estimate_prior_method)
+
+  # Initialize Model Object
+  model <- ibss_initialize(data                  = data,
+                           L                     = L,
+                           scaled_prior_variance = scaled_prior_variance,
+                           residual_variance     = residual_variance,
+                           prior_weights         = prior_weights,
+                           null_weight           = null_weight,
+                           model_init            = model_init)
+
+  # Initialize ELBO & Tracking
+  elbo <- rep(as.numeric(NA), max_iter + 1)
+  elbo[1] <- -Inf
+  tracking <- list()
+
+  # IBSS Loop
+  for(iter in seq_len(max_iter)){
+
+    # Save core values for each iteration
+
+    # This could be absorbed into ibss_fit, but i dont believe it should. I think
+    # it makes sense to absorb the various checks but when we are adding additional
+    # features to the model output it should remain more transparent.
+    tracking <- susie_extract_core(data, model, tracking, iter, track_fit)
+
+    # Update all L effects
+    model <- ibss_fit(data, model,
+                      estimate_prior_variance = estimate_prior_variance,
+                      estimate_prior_method   = estimate_prior_method,
+                      check_null_threshold    = check_null_threshold,
+                      check_prior             = check_prior)
+
+    # Get Objective & Check Convergence
+    elbo[iter + 1] <- get_objective(data, model)
+
+    if (elbo[iter + 1] - elbo[iter] < tol) {
+      model$converged <- TRUE
+      break
+    }
+
+    # Update Residual Variance
+    if (estimate_residual_variance) {
+      model$sigma2 <- max(residual_variance_lowerbound,
+                          est_residual_variance(data, model))
+      model$sigma2 <- min(model$sigma2, residual_variance_upperbound)
+    }
+  }
+
+  if (is.null(model$converged)) {
+    warning(paste("IBSS algorithm did not converge in",max_iter,"iterations!"))
+    model$converged = FALSE
+  }
+
+  model$elbo <- elbo[2:(iter + 1)]
+
+  rerun <- function(prior_weights, model_init = NULL) {
```

Comment:
Please make a note to discuss this `refine` function next time we meet -- I have some comments which I think is easier explain in person